pudb/var_view.py

```diff
@@ -386,84 +470,25 @@ def walk_value(self, parent, label, value, id_path=None, attr_prefix=None):
-                                cont_id_path)
-                            break
-
-                    self.walk_value(new_parent_item, None, entry,
-                        "%s[%d]" % (id_path, i))
-                if not value:
-                    self.add_item(new_parent_item, "<empty>", None)
-                return
-
             # containers --------------------------------------------------
-            key_it = None
-
-            try:
-                key_it = value.keys()
-            except AttributeError:
-                # keys or iterkeys doesn't exist, not worth mentioning!
-                pass
-            except Exception:
-                ui_log.exception("Failed to obtain key iterator")
-
-            if key_it is None:
-                try:
-                    len_value = len(value)
-                except (AttributeError, TypeError):
-                    # no __len__ defined on the value, not worth mentioning!
-                    pass
-                except Exception:
-                    ui_log.exception("Failed to determine container length")
-                else:
-                    try:
-                        value[0]
-                    except (LookupError, TypeError):
-                        key_it = []
-                    except Exception:
-                        ui_log.exception("Item is not iterable")
-                    else:
-                        key_it = xrange(len_value)
-
-            if key_it is not None:
-                cnt = 0
-                for key in key_it:
-                    if cnt % 10 == 0 and cnt:
-                        cont_id_path = "%s.cont-%d" % (id_path, cnt)
-                        if not self.frame_var_info.get_inspect_info(
-                                cont_id_path, read_only=True).show_detail:
-                            self.add_item(
-                                new_parent_item, "...", None, cont_id_path)
-                            break
-
-                    try:
-                        next_value = value[key]
-                    except (LookupError, TypeError):
-                        ui_log.exception("Failed to iterate an item that "
-                            "appeared to be iterable.")
-                        break
-
-                    self.walk_value(new_parent_item, repr(key), next_value,
-                        "%s[%r]" % (id_path, key))
-                    cnt += 1
-                if not cnt:
-                    self.add_item(new_parent_item, "<empty>", None)
+            if isinstance(value, PudbMapping):
+                self.walk_mapping(new_parent_item, label, value, id_path)
+                return
+            elif isinstance(value, PudbSequence):
+                self.walk_sequence(new_parent_item, label, value, id_path)
+                return
+            elif isinstance(value, PudbCollection):
+                self.walk_collection(new_parent_item, label, value, id_path)
```

Comment:
This question is not specific, but it's potentially worth asking: Is there a reason to worry about types that are one of these *and* have interesting attributes?

Reply:
Yeah I was wondering about this too actually. Even if we ditch the ABCs and use plain isinstance checks, this could still be a problem if someone, for example, subclasses `list` to add some interesting attributes.

Would it make sense to provide a "meta" attribute of some sort that displays contents, and expand the contents under that attribute?

Reply:
Actually another thought- this could apply to the "simple" types as well. Someone may well subclass `str` or `int` in some interesting way and want to inspect their added attributes.

Reply:
I like the idea of the "meta" attribute. That might not even be all that yucky in terms of the traversal logic.

Reply:
Okay I've pushed an adjustment that allows expansion on all variables, regardless if they have a "basic" type, are a container, or are some other object, and uses a "metaitem" to hold the contents of containers.

The tests will now need fixing up of course, so this push should fail, but since the nature of the tests will change based on how this is implemented, I would prefer to wait until we've settled on a final approach before fixing them.

Reply:
What do you think of the current approach to the "value string" for the contents meta-item? I might try putting together a "contents preview" since I'm not entirely happy with just showing "(hidden)".

Reply:
I decided to go ahead and implement the preview.